Improve output of failing tests comparing structs

`go-cmp`'s diffs leave a bit to be desired when debugging failing tests, this is to investigate using https://github.com/go-test/deep instead.

 `go-cmp.Diff` produces output that is pretty hard to read, for example:
```
--- FAIL: TestListenerVisit (0.00s)
    --- FAIL: TestListenerVisit/multiple_tls_ingress_with_secrets_should_be_sorted (0.00s)
        listener_test.go:631: {map[string]*v2.Listener}["ingress_https"]:
                -: s`name:"ingress_https" address:<socket_address:<address:"0.0.0.0" port_value:8443 > > filter_chains:<filter_chain_match:<server_names:"sortedfirst.example.com" > tls_context:<common_tls_context:<tls_params:<tls_minimum_protocol_version:TLSv1_1 tls_maximum_protocol_version:TLSv1_3 cipher_suites:"[ECDHE-ECDSA-AES128-GCM-SHA256|ECDHE-ECDSA-CHACHA20-POLY1305]" cipher_suites:"[ECDHE-RSA-AES128-GCM-SHA256|ECDHE-RSA-CHACHA20-POLY1305]" cipher_suites:"ECDHE-ECDSA-AES128-SHA" cipher_suites:"ECDHE-RSA-AES128-SHA" cipher_suites:"ECDHE-ECDSA-AES256-GCM-SHA384" cipher_suites:"ECDHE-RSA-AES256-GCM-SHA384" cipher_suites:"ECDHE-ECDSA-AES256-SHA" cipher_suites:"ECDHE-RSA-AES256-SHA" > tls_certificate_sds_secret_configs:<name:"default/secret/735ad571c1" sds_config:<api_config_source:<api_type:GRPC grpc_services:<envoy_grpc:<cluster_name:"contour" > > > > > alpn_protocols:"h2" alpn_protocols:"http/1.1" > > filters:<name:"envoy.http_connection_manager" typed_config:<type_url:"type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager" value:"\022\ringress_https\032\"\n\021\022\017\010\002\"\013\n\t\n\007contour\022\ringress_https*\014\n\nenvoy.gzip*\020\n\016envoy.grpc_web*\016\n\014envoy.routerB\002\020\001Z\002\010<je\n\025envoy.file_access_log\"L\n;type.googleapis.com/envoy.config.accesslog.v2.FileAccessLog\022\r\n\013/dev/stdoutr\002\010\001\362\001\002\010\001" > > > filter_chains:<filter_chain_match:<server_names:"sortedsecond.example.com" > tls_context:<common_tls_context:<tls_params:<tls_minimum_protocol_version:TLSv1_1 tls_maximum_protocol_version:TLSv1_3 cipher_suites:"[ECDHE-ECDSA-AES128-GCM-SHA256|ECDHE-ECDSA-CHACHA20-POLY1305]" cipher_suites:"[ECDHE-RSA-AES128-GCM-SHA256|ECDHE-RSA-CHACHA20-POLY1305]" cipher_suites:"ECDHE-ECDSA-AES128-SHA" cipher_suites:"ECDHE-RSA-AES128-SHA" cipher_suites:"ECDHE-ECDSA-AES256-GCM-SHA384" cipher_suites:"ECDHE-RSA-AES256-GCM-SHA384" cipher_suites:"ECDHE-ECDSA-AES256-SHA" cipher_suites:"ECDHE-RSA-AES256-SHA" > tls_certificate_sds_secret_configs:<name:"default/secret/735ad571c1" sds_config:<api_config_source:<api_type:GRPC grpc_services:<envoy_grpc:<cluster_name:"contour" > > > > > alpn_protocols:"h2" alpn_protocols:"http/1.1" > > filters:<name:"envoy.http_connection_manager" typed_config:<type_url:"type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager" value:"\022\ringress_https\032\"\n\021\022\017\010\002\"\013\n\t\n\007contour\022\ringress_https*\014\n\nenvoy.gzip*\020\n\016envoy.grpc_web*\016\n\014envoy.routerB\002\020\001Z\002\010<je\n\025envoy.file_access_log\"L\n;type.googleapis.com/envoy.config.accesslog.v2.FileAccessLog\022\r\n\013/dev/stdoutr\002\010\001\362\001\002\010\001" > > > listener_filters:<name:"envoy.listener.tls_inspector" > `
                +: s`name:"ingress_https" address:<socket_address:<address:"0.0.0.0" port_value:8443 > > filter_chains:<filter_chain_match:<server_names:"sortedsecond.example.com" > tls_context:<common_tls_context:<tls_params:<tls_minimum_protocol_version:TLSv1_1 tls_maximum_protocol_version:TLSv1_3 cipher_suites:"[ECDHE-ECDSA-AES128-GCM-SHA256|ECDHE-ECDSA-CHACHA20-POLY1305]" cipher_suites:"[ECDHE-RSA-AES128-GCM-SHA256|ECDHE-RSA-CHACHA20-POLY1305]" cipher_suites:"ECDHE-ECDSA-AES128-SHA" cipher_suites:"ECDHE-RSA-AES128-SHA" cipher_suites:"ECDHE-ECDSA-AES256-GCM-SHA384" cipher_suites:"ECDHE-RSA-AES256-GCM-SHA384" cipher_suites:"ECDHE-ECDSA-AES256-SHA" cipher_suites:"ECDHE-RSA-AES256-SHA" > tls_certificate_sds_secret_configs:<name:"default/sortedsecond-secret/735ad571c1" sds_config:<api_config_source:<api_type:GRPC grpc_services:<envoy_grpc:<cluster_name:"contour" > > > > > alpn_protocols:"h2" alpn_protocols:"http/1.1" > > filters:<name:"envoy.http_connection_manager" typed_config:<type_url:"type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager" value:"\022\ringress_https\032\"\n\021\022\017\010\002\"\013\n\t\n\007contour\022\ringress_https*\014\n\nenvoy.gzip*\020\n\016envoy.grpc_web*\016\n\014envoy.routerB\002\020\001Z\002\010<je\n\025envoy.file_access_log\"L\n;type.googleapis.com/envoy.config.accesslog.v2.FileAccessLog\022\r\n\013/dev/stdoutr\002\010\001\362\001\002\010\001" > > > filter_chains:<filter_chain_match:<server_names:"sortedfirst.example.com" > tls_context:<common_tls_context:<tls_params:<tls_minimum_protocol_version:TLSv1_1 tls_maximum_protocol_version:TLSv1_3 cipher_suites:"[ECDHE-ECDSA-AES128-GCM-SHA256|ECDHE-ECDSA-CHACHA20-POLY1305]" cipher_suites:"[ECDHE-RSA-AES128-GCM-SHA256|ECDHE-RSA-CHACHA20-POLY1305]" cipher_suites:"ECDHE-ECDSA-AES128-SHA" cipher_suites:"ECDHE-RSA-AES128-SHA" cipher_suites:"ECDHE-ECDSA-AES256-GCM-SHA384" cipher_suites:"ECDHE-RSA-AES256-GCM-SHA384" cipher_suites:"ECDHE-ECDSA-AES256-SHA" cipher_suites:"ECDHE-RSA-AES256-SHA" > tls_certificate_sds_secret_configs:<name:"default/sortedfirst-secret/735ad571c1" sds_config:<api_config_source:<api_type:GRPC grpc_services:<envoy_grpc:<cluster_name:"contour" > > > > > alpn_protocols:"h2" alpn_protocols:"http/1.1" > > filters:<name:"envoy.http_connection_manager" typed_config:<type_url:"type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager" value:"\022\ringress_https\032\"\n\021\022\017\010\002\"\013\n\t\n\007contour\022\ringress_https*\014\n\nenvoy.gzip*\020\n\016envoy.grpc_web*\016\n\014envoy.routerB\002\020\001Z\002\010<je\n\025envoy.file_access_log\"L\n;type.googleapis.com/envoy.config.accesslog.v2.FileAccessLog\022\r\n\013/dev/stdoutr\002\010\001\362\001\002\010\001" > > > listener_filters:<name:"envoy.listener.tls_inspector" > `
```

When run against the same failing test, `deep.Equal` produces this:
```
--- FAIL: TestListenerVisit (0.00s)
    --- FAIL: TestListenerVisit/multiple_tls_ingress_with_secrets_should_be_sorted (0.00s)
        listener_test.go:630: [map[ingress_https].FilterChains.slice[0].FilterChainMatch.ServerNames.slice[0]: sortedfirst.example.com != sortedsecond.example.com map[ingress_https].FilterChains.slice[1].FilterChainMatch.ServerNames.slice[0]: sortedsecond.example.com != sortedfirst.example.com]
```
Much easier to read.

Currently, though, `deep.Equal` does not compare unexported fields, so we need to check if that's possible. If not, it may, sadly, not be a candidate.

--- a/internal/assert/assert.go
+++ b/internal/assert/assert.go
@@ -29,6 +29,9 @@ type Assert struct {
 	t *testing.T
 }
 
+// Equal will test that want == got, and call t.Fatal if it does not.
+// Notably, for errors, they are equal if they are both nil, or are both non-nil.
+// No value information is checked for errors.
 func Equal(t *testing.T, want, got interface{}) {
 	t.Helper()
 	Assert{t}.Equal(want, got)

